File: 142-linked-list-cycle-ii/142-linked-list-cycle-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        if (!head)
            return nullptr;
        
        ListNode *fast = head;
        ListNode *slow = head;
        
        while (fast->next && fast->next->next) {
            fast = fast->next->next;
            slow = slow->next;
            
            if (slow == fast) {
                ListNode *third = head;
                while (slow != third) {
                    slow = slow->next;
                    third = third->next;
                }
                return third;
            }
        }
        
        return nullptr;
    }
        
};

File: 142-linked-list-cycle-ii/142-linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head is None:
            return None
        
        fast = head
        slow = head
        
        while fast.next is not None and fast.next.next is not None:
            fast = fast.next.next
            slow = slow.next
            
            if fast == slow:
                third = head
                while third != slow:
                    third = third.next
                    slow = slow.next
                return third
            
        return None
        